active/0000-trait-based-exception-handling.md

```diff
@@ -0,0 +1,748 @@
+of other possibilities, each with different advantages and drawbacks:
+
+ * `try { ... } catch { ... }`
+
+ * `try { ... } match { ... }`
+
+ * `try { ... } handle { ... }`
+
+ * `catch { ... } match { ... }`
+
+ * `catch { ... } handle { ... }`
+
+ * `catch ...` (without braces or a second clause)
+
+Among the considerations:
+
+ * Simplicity. Brevity.
+
+ * Following precedent from existing, popular languages, and familiarity with
+   respect to analogous constructs in them.
+
+ * Fidelity to the constructs' actual behavior. For instance, the first clause
+   always catches the "exception"; the second only branches on it.
+
+ * Consistency with the existing `try!()` macro. If the first clause is called
+   `try`, then `try { }` and `try!()` would have essentially inverse meanings.
+
+ * Language-level backwards compatibility when adding new keywords. I'm not sure
+   how this could or should be handled.
+
+
+## Semantics for "upcasting"
+
+What should the contract for a `From`/`Into` `impl` be? Are these even the right
+`trait`s to use for this feature?
+
+Two obvious, minimal requirements are:
+
+ * It should be pure: no side effects, and no observation of side effects. (The
+   result should depend *only* on the argument.)
+
+ * It should be total: no panics or other divergence, except perhaps in the case
+   of resource exhaustion (OOM, stack overflow).
+
+The other requirements for an implicit conversion to be well-behaved in the
+context of this feature should be thought through with care.
+
+Some further thoughts and possibilities on this matter:
+
+ * It should be "like a coercion from subtype to supertype", as described
+   earlier. The precise meaning of this is not obvious.
+
+ * A common condition on subtyping coercions is coherence: if you can
+   compound-coerce to go from `A` to `Z` indirectly along multiple different
+   paths, they should all have the same end result.
+
+ * It should be unambiguous, or preserve the meaning of the input:
+   `impl From<u8> for u32` as `x as u32` feels right; as `(x as u32) * 12345`
+   feels wrong, even though this is perfectly pure, total, and injective. What
+   this means precisely in the general case is unclear.
+
+ * It should be lossless, or in other words, injective: it should map each
+   observably-different element of the input type to observably-different
+   elements of the output type. (Observably-different means that it is possible
+   to write a program which behaves differently depending on which one it gets,
+   modulo things that "shouldn't count" like observing execution time or
+   resource usage.)
+
+ * The types converted between should the "same kind of thing": for instance,
+   the *existing* `impl From<u32> for Ipv4Addr` is pretty suspect on this count.
+   (This perhaps ties into the subtyping angle: `Ipv4Addr` is clearly not a
+   supertype of `u32`.)
+
+
+## Forwards-compatibility
+
+If we later want to generalize this feature to other types such as `Option`, as
```

Comment:
Why? I mentioned HKT under Drawbacks and Alternatives, but the two features don't really interact, so compatibility isn't a concern.